### PR TITLE
Add more comprehensive tests for deserializing records

### PR DESCRIPTION
**Pull request by cmathiesen**
Adding some more SerDe tests to check that records including a map and list type deserialize properly, as well as started testing some more complex schemas/records

### CODE REVIEW

File: src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java [hunk 1]
```diff
@@ -84,7 +89,95 @@ public void testDeserializer() throws IOException, SerDeException {
...
+    List<?> deserialized = (List<?>) serde.deserialize(value);
+
+    assertEquals("Michael", deserialized.get(0));
+    assertEquals(3000L, deserialized.get(1));
+  }
+
+  @Test
+  public void testDeserializeMap() {
+    Schema schema = new Schema(required(1, "map_type", Types.MapType
+        .ofRequired(18, 19, Types.StringType.get(), Types.StringType.get())));
```
Comment (nvitucci) [inline src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java:101]: What is the `keyId` here? Is it the `TypeID` enum value or a "random" ID?

**Review by nvitucci — COMMENTED** [review]
It looks good to me. Just a comment for clarification.

File: src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java (hunk 1 above)
Comment (cmathiesen) [reply src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java:101]: It's a "random" value you choose - each field in the schema has an ID unique to that schema. I believe it's used to allow schema evolution: https://iceberg.incubator.apache.org/evolution/#correctness

File: src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java [hunk 2]
```diff
@@ -84,7 +89,95 @@ public void testDeserializer() throws IOException, SerDeException {
     reader.next(null, value);
 
     IcebergSerDe serde = new IcebergSerDe();
-    Object row = serde.deserialize(value);
-    //TODO: Finish this test with assertions
+    List<?> deserialized = (List<?>) serde.deserialize(value);
```
Comment (massdosage) [inline src/test/java/com/expediagroup/hiveberg/TestIcebergSerDe.java]: I'm not sure the `?` adds anything here. Could you just use `Object` instead? Or does that mean you need to do additional casting below?

**Review by massdosage — COMMENTED** [review]
I think I don't understand the contract of the SerDe - where do the values that are deserialized get used? I was expecting this to return Records but I guess it needs to return the contents of the records so that Hive can use them?

**Comment by cmathiesen** [comment]
> I think I don't understand the contract of the SerDe - where do the values that are deserialized get used? I was expecting this to return Records but I guess it needs to return the contents of the records so that Hive can use them?

From Hive docs: `HDFS files --> InputFileFormat --> <key, value> --> Deserializer --> Row object`

So from my understanding, the deserialize method extracts all the separate objects that are part of a single row (which is usually one long string or some binary) into separate "columns" which we can store simply in a list. I investigated this after examining [https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/avro/AvroDeserializer.java#L161](https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/avro/AvroDeserializer.java#L161).  Then Hive uses the ObjectInspector to get the actual field value out of each field we've deserialized. I think the SerDe is an area that could use more investigation because what I've written is super basic and maybe haven't understood all the complexities fully

**Review by nvitucci — APPROVED** [review]

**Comment by massdosage** [comment]
> > I think I don't understand the contract of the SerDe - where do the values that are deserialized get used? I was expecting this to return Records but I guess it needs to return the contents of the records so that Hive can use them?
> 
> From Hive docs: `HDFS files --> InputFileFormat --> <key, value> --> Deserializer --> Row object`
> 
> So from my understanding, the deserialize method extracts all the separate objects that are part of a single row (which is usually one long string or some binary) into separate "columns" which we can store simply in a list. I investigated this after examining https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/avro/AvroDeserializer.java#L161. Then Hive uses the ObjectInspector to get the actual field value out of each field we've deserialized. I think the SerDe is an area that could use more investigation because what I've written is super basic and maybe haven't understood all the complexities fully

OK, that makes sense, and yeah, the Avro one is probably a good reference implementation to follow.

**Review by massdosage — APPROVED** [review]

**Pull request merged**